src/random_ray/flat_source_domain.cpp

```diff
@@ -68,6 +78,25 @@ FlatSourceDomain::FlatSourceDomain() : negroups_(data::mg.num_energy_groups_)
   if (source_region_id != n_source_regions_) {
     fatal_error("Unexpected number of source regions");
   }
+
+  // Initialize tally volumes
+  tally_volumes_.resize(model::tallies.size());
+  for (int i = 0; i < model::tallies.size(); i++) {
+    //  Get the shape of the 3D result tensor
+    auto shape = model::tallies[i]->results().shape();
+
+    // Create a new 2D tensor with the same size as the first
+    // two dimensions of the 3D tensor
+    tally_volumes_[i] =
+      xt::xtensor<double, 2>::from_shape({shape[0], shape[1]});
+  }
+
+  // Compute simulation domain volume based on ray source
+  auto* is = dynamic_cast<IndependentSource*>(RandomRay::ray_source_.get());
```

Comment:
I recommend checking that `is` is not null here. The check may happen elsewhere too, but it's good defensive programming. I wish we used `assert` from `cassert` because it'd be perfect for this kind of thing. A simple `assert(is)` would be great.

And it's nice that this is a no-op if compiled with `-DNDEBUG`.